### PR TITLE
compute single loss in sgd, move prediction code to devicde function,…

### PR DESCRIPTION
**Pull request by nickgreenquist**
… add biases to shared mem, add timers to training loop and total loss calcs

### CODE REVIEW

File: matrix_factorization/loss.cu [hunk 1]
```diff
@@ -9,30 +9,35 @@
 #include "matrix.h"
 
 #define index(i, j, N)  ((i)*(N)) + (j)
+#define warp_size 32 //TODO: we need to get device props
 
 using namespace cu2rec;
 
 // PARALLEL
+extern __shared__ float biases[];
```
Comment (dorukkilitcioglu) [inline matrix_factorization/loss.cu]: Does this need to be outside? Can probably go inside the `loss_kernel`.

File: matrix_factorization/loss.cu [hunk 2]
```diff
@@ -9,30 +9,35 @@
...
 __global__ void loss_kernel(int factors, int user_count, int item_count, const float * P, const float * Q, const int * indptr, 
                             const int * indices, const float * data, float * error, float * user_bias, float * item_bias, float global_bias) {
+
+    // use first warp to load in user_biases
+    float* s_user_bias = (float*)biases;
+    if(threadIdx.x < warp_size) {
+        for(int i = 0; i < user_count; i += warp_size) {
```
Comment (dorukkilitcioglu) [inline matrix_factorization/loss.cu]: So, this doesn't use anything related to current thread at all, meaning this gets repeated for all threads. Which also makes me think that only the first user of each warp is loaded into the shared memory. Instead, maybe we can let all threads load in their own biases to shared mem?

File: matrix_factorization/sgd.cu [hunk 3]
```diff
@@ -16,41 +17,52 @@ __global__ void initCurand(curandState *state, unsigned long seed, int n_rows){
     }
 }
 
-__global__ void sgd_update(int *indptr, int *indices, float *P, float *Q, float *P_target, float *Q_target, 
+extern __shared__ float biases[];
```
Comment (dorukkilitcioglu) [inline matrix_factorization/sgd.cu]: Again, can't this go inside the function?

File: matrix_factorization/sgd.cu [hunk 4]
```diff
@@ -16,41 +17,52 @@ __global__ void initCurand(curandState *state, unsigned long seed, int n_rows){
...
                            float *errors, int n_rows, int n_cols, float *user_bias, float *item_bias,
-                           float *user_bias_target, float *item_bias_target, curandState *my_curandstate) {
+                           float *item_bias_target, curandState *my_curandstate,
+                            float global_bias) {
+
+    // use first warp to load in user_biases
+    float* s_user_bias = (float*)biases;
+    if(threadIdx.x < warp_size) {
+        for(int i = 0; i < n_rows; i += warp_size) {
+            s_user_bias[i] = user_bias[i];
```
Comment (dorukkilitcioglu) [inline matrix_factorization/sgd.cu]: Again, not sure if this is loading everything correctly.

File: matrix_factorization/sgd.cu [hunk 5]
```diff
@@ -16,41 +17,52 @@ __global__ void initCurand(curandState *state, unsigned long seed, int n_rows){
...
+
         int y = indices[y_i];
         for(int f = 0; f < config::n_factors; ++f) {
             int p_index = index(x, f, config::n_factors);
             int q_index = index(y, f, config::n_factors);
 
-            // Update P
-            P_target[p_index] = P[p_index] + config::learning_rate * (errors[y_i] * Q[q_index] - config::P_reg * P[p_index]);
-
-            // Only update Q if train flag is true
```
Comment (dorukkilitcioglu) [inline matrix_factorization/sgd.cu]: Can we add back this comment

File: matrix_factorization/training.cu [hunk 6]
```diff
@@ -167,16 +186,19 @@ void train(CudaCSRMatrix* train_matrix, CudaCSRMatrix* test_matrix, config::Conf
...
+        // Swap item related components
         swap(Q_device, Q_device_target);
-
-        // Swap old and new bias arrays
-        swap(user_bias_device, user_bias_target);
         swap(item_bias_device, item_bias_target);
 
         cfg->cur_iterations += 1;
+
+        cudaDeviceSynchronize();
```
Comment (dorukkilitcioglu) [inline matrix_factorization/training.cu]: Do we need this to be inside the for loop?

File: matrix_factorization/loss.cu (hunk 1 above)
Comment (nickgreenquist) [reply matrix_factorization/loss.cu]: it did when there were two shared mem arrays, now there are one. So yeah, we can put it back inside
